### DISCUSSION TITLE
TS0201, _TZ3210_ncw88jfq, LCD Temperature Unit Change

### DISCUSSION THREAD

**Discussion by mag2352** [Koenkk/zigbee2mqtt#23345] [Devices]
I have a TS0201, which is a temperature/humidity sensor with a screen. It is mostly supported by z2m, but also has the ability to change the units from C to F on the screen from the Tuya app. This functionality isn't in z2m. The DP in the Tuya IoT website is shown as:

```
{
        "code": "lcd_temp_unit",
        "custom_name": "",
        "dp_id": 101,
        "time": 1720903782306,
        "type": "enum",
        "value": "f"
}
```

I tried making an external converter to make this feature work, but I haven't had success. I just set it as a temperatureunitenum, but I'm not sure if that's the right idea.

```
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const modernExtend = require('zigbee-herdsman-converters/lib/modernExtend');
const e = exposes.presets;
const ea = exposes.access;
const tuya = require('zigbee-herdsman-converters/lib/tuya');

const definition = {
    // Since a lot of Tuya devices use the same modelID, but use different datapoints
    // it's necessary to provide a fingerprint instead of a zigbeeModel
    fingerprint: [
        {
            // The model ID from: Device with modelID 'TS0601' is not supported
            // You may need to add \u0000 at the end of the name in some cases
            modelID: 'TS0201',
            // The manufacturer name from: Device with modelID 'TS0601' is not supported.
            manufacturerName: '_TZ3210_ncw88jfq',
        },
    ],
    model: 'TS0201_new',
    vendor: 'Tuya',
    description: 'Fill in a description of the device here',
    fromZigbee: [tuya.fz.datapoints],
    toZigbee: [tuya.tz.datapoints],
    onEvent: tuya.onEventSetTime, // Add this if you are getting no converter for 'commandMcuSyncTime'
    configure: tuya.configureMagicPacket,
    exposes: [
        // Here you should put all functionality that your device exposes
        e.temperature(),
        e.humidity(),
        e.battery(),
        tuya.exposes.temperatureUnit()
    ],
    meta: {
        // All datapoints go in here
        tuyaDatapoints: [
            [1, 'temperature', tuya.valueConverter.raw],
            [2, 'humidity', tuya.valueConverter.raw],
            [4, 'battery', tuya.valueConverter.raw]
            [101, 'temperature_unit', tuya.valueConverter.temperatureUnitEnum],
        ],
    },
    extend: [
        // A preferred new way of extending functionality.
    ],
};

module.exports = definition;
```

When I try to set the unit in z2m after using the external converter, this message appears in the logs:

```
[2024-07-15 15:54:48] debug: 	z2m:mqtt: Received MQTT message on 'zigbee2mqtt/Tuya TH Screen/set' with data '{"temperature_unit":"c"}'
[2024-07-15 15:54:48] debug: 	z2m: Publishing 'set' 'temperature_unit' to 'Tuya TH Screen'
[2024-07-15 15:54:48] error: 	z2m: Publish 'set' 'temperature_unit' to 'Tuya TH Screen' failed: 'TypeError: Cannot read properties of undefined (reading '1')'
[2024-07-15 15:54:48] debug: 	z2m: TypeError: Cannot read properties of undefined (reading '1')
```

Am I doing something wrong? Or is this something that cannot be realistically implemented?

**Comment by czsmith** [comment]
Any updates on this issue?  I'd like to do the same thing as well.

**Comment by fresnoboy** [comment]
+1  Yes, I am having this issue too.  Does it stick if I reset it in the Tuya app, and then I pair it to z2m?

**Reply by TheNutHouse84** [reply]
No, tried multiple times pairing to tuya hub, setting to F in tuya app (sensor display changes) then pairing to Z2M. Initiating the pairing process resets the sensor to factory.

**Comment by jspiros** [comment]
I figured out how to do it from the dev console, and wrote a note, thinking I'd eventually figure out how to contribute it to the converter z2m uses, but haven't had the time. So, maybe someone else can figure it out from this before I ever get around to it.

This is how I switched from Celsius to Fahrenheit using the Z2M dev console:

> endpoint 1 cluster manuSpecificTuya (0xef00) command sendData (0x04) payload
{"dpValues":[{"data":"1","datatype":4,"dp":101}],"seq":1337}
> 
> (seq arbitrary/context sensitive, 1337 just to have something worked)

If data is set to 0 instead of 1, it switches back to Celsius from Fahrenheit.

It is a sleepy device, so IIRC I had to press the button while executing the command.

**Reply by jspiros** [reply]
Retracing my steps, I believe this was where I got the information I needed: https://github.com/kkossev/Hubitat/blob/8931b9ccbabdc180c36016badb9791b491d3ca40/Drivers/Tuya%20Temperature%20Humidity%20Illuminance%20LCD%20Display%20with%20a%20Clock/Tuya_Temperature_Humidity_Illuminance_LCD_Display_with_a_Clock.groovy#L1162

**Reply by jspiros** [reply]
And the author of that credited this person, I believe for this post, or at least that thread leading up to this post: https://community.hubitat.com/t/release-tuya-temperature-humidity-illuminance-lcd-display-with-a-clock-w-healthstatus/88093/581
